File: frontend/src/app/core/services/usuario.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Usuario, UsuarioRequest } from '../../models/usuario.model';

@Injectable({ providedIn: 'root' })
export class UsuarioService {

  private readonly API_URL = '/api/usuarios';

  constructor(private http: HttpClient) {}

  listarTodos(): Observable<Usuario[]> {
    return this.http.get<Usuario[]>(this.API_URL);
  }

  buscarPorId(id: number): Observable<Usuario> {
    return this.http.get<Usuario>(`${this.API_URL}/${id}`);
  }

  crear(dto: UsuarioRequest): Observable<Usuario> {
    return this.http.post<Usuario>(this.API_URL, dto);
  }

  actualizar(id: number, dto: UsuarioRequest): Observable<Usuario> {
    return this.http.put<Usuario>(`${this.API_URL}/${id}`, dto);
  }

  cambiarEstado(id: number, activo: boolean): Observable<void> {
    return this.http.patch<void>(`${this.API_URL}/${id}/estado`, null, { params: { activo } });
  }

  eliminar(id: number): Observable<void> {
    return this.http.delete<void>(`${this.API_URL}/${id}`);
  }

  buscarPorUsername(username: string): Observable<Usuario> {
    return this.http.get<Usuario>(`${this.API_URL}/me`);
  }

  cambiarPassword(id: number, dto: { passwordActual: string; passwordNuevo: string }): Observable<void> {
    return this.http.put<void>(`${this.API_URL}/me/password`, dto);
  }
}

File: backend/src/main/java/com/kineticrehab/service/UsuarioService.java
package com.kineticrehab.service;

import com.kineticrehab.dto.request.CambioPasswordRequestDTO;
import com.kineticrehab.dto.request.UsuarioRequestDTO;
import com.kineticrehab.dto.response.UsuarioResponseDTO;

import java.util.List;

public interface UsuarioService {

    List<UsuarioResponseDTO> listarTodos();

    UsuarioResponseDTO buscarPorId(Long id);

    UsuarioResponseDTO crear(UsuarioRequestDTO dto);

    UsuarioResponseDTO actualizar(Long id, UsuarioRequestDTO dto);

    void cambiarEstado(Long id, Boolean activo);

    void eliminar(Long id);

    UsuarioResponseDTO buscarPorUsername(String username);

    void cambiarPassword(Long id, CambioPasswordRequestDTO dto);
}
